tests/unit/test_charm.py

```diff
@@ -6,39 +6,70 @@
 # Need access to protected functions for testing
 # pylint:disable=protected-access
 
+import secrets
 from functools import partial
 from typing import Any, Callable, cast
 from unittest.mock import MagicMock
 
+import jenkinsapi
 import pytest
 import requests
-from ops.charm import PebbleReadyEvent
+from ops.charm import ActionEvent, PebbleReadyEvent
 from ops.model import ActiveStatus, BlockedStatus, StatusBase
-from ops.testing import Harness
 
-import jenkins as jenkins_src
+import charm
+import jenkins
 from charm import JenkinsK8SOperatorCharm
 
 from .types_ import HarnessWithContainer
 
+ACTIVE_STATUS_NAME = "active"
+BLOCKED_STATUS_NAME = "blocked"
+MAINTENANCE_STATUS_NAME = "maintenance"
 
-def test__on_jenkins_pebble_ready_no_container(harness: Harness):
+
+def test__on_jenkins_pebble_ready_no_container(harness_container: HarnessWithContainer):
     """
     arrange: given a pebble ready event with container unable to connect.
     act: when the Jenkins pebble ready event is fired.
     assert: the event should be deferred.
     """
-    harness.set_can_connect(harness.model.unit.containers["jenkins"], False)
-    harness.begin()
-    charm = cast(JenkinsK8SOperatorCharm, harness.charm)
+    harness_container.harness.begin()
+    jenkins_charm = cast(JenkinsK8SOperatorCharm, harness_container.harness.charm)
     mock_event = MagicMock(spec=PebbleReadyEvent)
     mock_event.workload = None
 
-    charm._on_jenkins_pebble_ready(mock_event)
+    jenkins_charm._on_jenkins_pebble_ready(mock_event)
 
     mock_event.defer.assert_called()
 
 
+def test__on_jenkins_pebble_ready_error(
+    harness_container: HarnessWithContainer,
+    mocked_get_request: Callable[[str, int, Any, Any], requests.Response],
+    monkeypatch: pytest.MonkeyPatch,
+    raise_exception: Callable,
+):
+    """
+    arrange: given a patched jenkins bootstrap method that raises an exception.
+    act: when the jenkins_pebble_ready event is fired.
+    assert: the unit status should be in BlockedStatus.
+    """
+    # speed up waiting by changing default argument values
+    monkeypatch.setattr(jenkins.wait_ready, "__defaults__", (1, 1))
```

Comment:
This still doesn't seem optimal, not in this PR, in the background perhaps think about how this could be done differently?